WIP: refactor dimension scales handling

kmuehlbauer opened the pull request:
Fixes #88 

UPDATE: This PR was split into several smaller ones, it will be kept here for reference until completely resolved.

- Dimension scales are now created and attached on the fly, instead of doing the iteration on `flush` over all groups/subgroups.
- When opening a file for writing the maximum dimension id (`_Netcdf4Dimid`) is determined and increased for every newly created dimension (#104)
- Transparent handling of `_nc4_non_coord` (#102)
- Refactored out several functions for dim_scale handling at `h5netcdf.Group` and `h5netcdf.Variable` level
- Automatic resizing of variables and attached unlimited dimensions when writing to variables (#103)
- Adapted `resize_dimension` for both dimension resizing and recursive/non-recursive variable resizing (#103)
- Refactored out `_create_h5netcdf_variable`-function which creates hdf5 dataset and `h5netcdf.Variable` (#105)
- Simplified `_create_child_variable`-function with added checks which (netcdf) variable type will be created (#105) 

Bullet point 2 is now the last bottleneck since we need to iterate over all datasets to get the dimension id. I've used `visititems()` for the iteration.  There is room for further refactoring.

Todo:
- [x] simplifiy variable type checks
- [x] add/update tests for resizing dimensions/variables
- [x] add tests for dimension scale handling (eg. attaching/detaching dimension scales)
- [x] add tests for _nc4_non_coord_ handling
- [x] add/update tests for variable creation
- [ ] add more tests

For performance testing I used the code provided by @aldanor:
```python
import time
import xarray as xr
import numpy as np
import h5py

arr = xr.DataArray(np.random.RandomState(0).randint(-100, 100, (50_000, 3)), dims=['x', 'y'])
ds = xr.Dataset({'arr': arr})

filename = 'test.h5'
save = lambda group: ds.to_netcdf(filename, engine='h5netcdf', mode='a', group=str(group))

with h5py.File(filename, 'w') as _:
    pass

for i in range(250):
    save(i)
```
|  netcdf4    |  h5netcdf 0.11.0  |  h5netcdf-refactor |
|-----|------|-----|
| 15 sec   | 208 sec | 20 sec |

So this is a great improvement, although we do not match `netCDF4`.

Some valuable links on dimension scales:
- https://www.unidata.ucar.edu/blogs/developer/en/entry/dimensions_scales
- https://www.unidata.ucar.edu/blogs/developer/en/entry/dimension_scale2
- https://www.unidata.ucar.edu/blogs/developer/en/entry/dimension_scales_part_3
- https://www.unidata.ucar.edu/blogs/developer/en/entry/netcdf4_shared_dimensions
- https://www.unidata.ucar.edu/blogs/developer/en/entry/netcdf4_use_of_dimension_scales
- https://www.researchgate.net/publication/330347054_2A5_NETCDF-4_PERFORMANCE_IMPROVEMENTS_OPENING_COMPLEX_DATA_FILES

String `NULLTERM` vs `NULLPAD`:
- https://github.com/PyTables/PyTables/issues/264

Testing Notebook:
- https://gist.github.com/kmuehlbauer/99ad4749c4e5923b1364187c6363930a

netcdf4-python quirks:
- `_Netcdf4Dimid` gets attached to all data variables if a 2D coordinate variable is created  and any variable is written/file is reopened for append, see https://github.com/Unidata/netcdf4-python/issues/1104
- unlimited variable dimensions are reported as current size of the dimension scale, even if the variable's underlying `DATASPACE` dimension is smaller (eg.0)

groutr commented on h5netcdf/core.py:
```diff
@@ -726,6 +747,7 @@ def __init__(self, path, mode="a", invalid_netcdf=None, phony_dims=None, **kwarg
             self._closed = False
 
         self._mode = mode
+        self._writable = not (mode == "r")
```
Would this be the same as `mode != 'r'`?

kmuehlbauer replied:
Now that you mention this...This looks much nicer. Why did I do this in the first place?

shoyer commented on h5netcdf/core.py:
```diff
@@ -674,6 +690,11 @@ def resize_dimension(self, dimension, size):
         # Resize the dimension.
         self._current_dim_sizes[dimension] = size
 
+        # If not a variable but dimension -> resize
+        if dimension not in self.variables and dimension in self._h5group:
+            self._h5group[dimension].resize((size,))
```
This looks sensible, but is it related somehow to the other changes? It might warrant another unit-test.

kmuehlbauer replied:
If only a dimension is created, the corresponding Dataset will be created at the same time. Before it would only be created on `flush`.

Now, when we create a dimension and resize it afterwards, we have to resize here since the Dataset already exists. 

I'll take a look and come up with some more tests with this. Good chance that this will lead to some more refactoring.

kmuehlbauer replied:
@shoyer I've added tests, which checks correct resizing of dimensions vs. variables.

shoyer reviewed (commented):
what a nice optimization! 😍

kmuehlbauer commented:
OK; I've found some inconsistencies with the current code, the refactored code and the netcdf behaviour. I'll need to dive a bit deeper.

kmuehlbauer commented:
@shoyer I've tried hard to break this into pieces, but it's too much functionally interdepending.

I've worked on these things:

- transparent handling of `_nc4_non_coord`
- automatic resizing of variables and attached unlimited dimensions when writing to variables
- refactored out `_create_h5netcdf_variable`-function which creates hdf5 dataset and `h5netcdf.Variable`
- simplified `_create_child_variable`-function with added checks which (netcdf) variable type will be created (these checks need another refactor, since the boolean logic hard to read atm)
- refactored out several functions for dim_scale handling at `h5netcdf.Group` and `h5netcdf.Variable` level
- dim_scales are created/removed and attached/detached on the fly
- adapted `resize_dimension` for both dimension resizing and recursive/non-recursive variable resizing

I did not touch the tests yet, since I wanted to be fully backwards compatible regarding existing tests. That all tests run is good indicator that the refactor was successful. I've also checked the details (netCDF4, h5netcdf, legacyapi) inside a jupyter notebook using kind of read/write matrix:

- write netCDF4, write legacyapi, write h5netcdf
- read netCDF4, read legacyapi, read h5netcdf
- checking `ncdump` and `h5dump` on all created files, diffing the output to check for consistency

I'll make the notebook available, after a clean-up. Tests follow too. Looking forward to hear your thoughts.

kmuehlbauer commented:
@shoyer I've added a link to the testing notebook in the OP.

kmuehlbauer commented:
@shoyer This is ready from my side for review.

shoyer commented:
With regards to the test coverage -- right now, it looks like you are mostly testing h5netcdf only private APIs. This is a fine start, but would be ideal to also test public interfaces, and to verify that netCDF4 creates equivalent HDF5 files.

kmuehlbauer commented on h5netcdf/core.py:
```diff
@@ -73,6 +72,66 @@ def _invalid_netcdf_feature(feature, allow, file, stacklevel=0):
+    number of full slices and then padding the key with full slices so that it
+    reaches the appropriate dimensionality.
+    """
+    # Taken from xarray:
+    # https://github.com/pydata/xarray/blob/835a53e62b1e5018faa323c649149e8294dd6af7/xarray/core/indexing.py#L23
```
@shoyer Thanks, I'll add it.

kmuehlbauer commented:
So the round-trip tests are not sufficient? What do you suggest to test this?

kmuehlbauer commented:
I can take the example from the Unidata blog and enhance it a bit.

shoyer commented:
> So the round-trip tests are not sufficient? What do you suggest to test this?

Well, we do still have the existing checks. And if this change is only speeding things up (not changing functionality), then they are probably fine.

kmuehlbauer commented:
@shoyer It changes functionality somehow in that sense that every change to dimensions will instantly be applied to the underlying datasets and not on `close()`/`flush()`. That has effects when variables are created and written. I'm trying to address everything but it can't be split into pieces because of the involved machinery. With every step forward I encounter another inconsistency between current h5netcdf behaviour and netCDF4.

I'll make this WIP again. This needs much more love.

kmuehlbauer commented:
@shoyer One example of those inconsistencies: https://github.com/Unidata/netcdf4-python/issues/1104

kmuehlbauer commented:
This PR has been superseded by #102, #103, #104 and #112. I'll close this. Refernces in the first post will be moved to the Wiki.
